File: fixtures/scale/dart/large/lib/ciphers/gronsfeld/gronsfeld_cipher.dart
import '../../core/cipher.dart';
import '../../util/alphabet.dart';
import 'gronsfeld_key.dart';

/// Vigenere restricted to digit keys: each digit is a shift.
class GronsfeldCipher implements Cipher {
  final GronsfeldKey key;

  GronsfeldCipher(this.key);

  @override
  String name() => 'gronsfeld';

  @override
  String encrypt(String plaintext) => _transform(cleanAlphabet(plaintext), true);

  @override
  String decrypt(String ciphertext) => _transform(cleanAlphabet(ciphertext), false);

  String _transform(String text, bool forward) {
    final sb = StringBuffer();
    for (var i = 0; i < text.length; i++) {
      final x = indexOf(text[i]);
      final k = key.digitAt(i);
      sb.write(charAt(forward ? x + k : x - k));
    }
    return sb.toString();
  }
}

File: fixtures/scale/kotlin/large/src/cryptobench/ciphers/gronsfeld/GronsfeldCipher.kt
package cryptobench.ciphers.gronsfeld

import cryptobench.core.Cipher
import cryptobench.util.Alphabet

/** Vigenere restricted to digit keys: each digit is a shift. */
class GronsfeldCipher(private val key: GronsfeldKey) : Cipher {

    override fun name(): String = "gronsfeld"

    override fun encrypt(plaintext: String): String = transform(Alphabet.clean(plaintext), true)

    override fun decrypt(ciphertext: String): String = transform(Alphabet.clean(ciphertext), false)

    private fun transform(text: String, forward: Boolean): String {
        val sb = StringBuilder()
        for (i in text.indices) {
            val x = Alphabet.indexOf(text[i])
            val k = key.digitAt(i)
            sb.append(Alphabet.charAt(if (forward) x + k else x - k))
        }
        return sb.toString()
    }
}
